File: app/src/main/java/com/alex/materialdiary/sys/adapters/RecycleAdapterUsers.java
package com.alex.materialdiary.sys.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.RecyclerView;

import com.alex.materialdiary.ChangeUserFragment;
import com.alex.materialdiary.R;
import com.alex.materialdiary.ShareQRFragmentDirections;
import com.alex.materialdiary.WebLoginFragmentDirections;
import com.alex.materialdiary.sys.ImageLoader;
import com.alex.materialdiary.sys.common.CommonAPI;
import com.alex.materialdiary.sys.common.models.ShareUser;
import com.alex.materialdiary.sys.common.models.get_user.Participant;
import com.alex.materialdiary.sys.common.models.get_user.SchoolInfo;
import com.alex.materialdiary.sys.common.models.get_user.Schools;
import com.alex.materialdiary.sys.common.models.period_marks.Mark;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class RecycleAdapterUsers extends RecyclerView.Adapter<RecycleAdapterUsers.ViewHolder>{

    private final LayoutInflater inflater;
    private final ChangeUserFragment fragment;
    private final List<Participant> participants;
    private final List<SchoolInfo> schoolsForUsers = new ArrayList<>();

    public RecycleAdapterUsers(ChangeUserFragment context, List<Schools> schools) {
        this.inflater = LayoutInflater.from(context.requireContext());
        participants = transform(schools);
        fragment = context;
    }

    @NonNull
    @Override
    public RecycleAdapterUsers.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.users_item, parent, false);
        return new ViewHolder(view);
    }
    List<Participant> transform(List<Schools> schools){
        List<Participant> list = new ArrayList<Participant>();
        if (schools == null) return null;
        for (Schools sch: schools) {
            if (sch.getRoles().contains("participant")) {
                list.add(sch.getParticipant());
                schoolsForUsers.add(sch.getSchool());
            }
            else {
                list.addAll(sch.getUserParticipants());
                for (int i = 0; i < sch.getUserParticipants().size(); i++) {
                    schoolsForUsers.add(sch.getSchool());
                }
            }
        }
        return list;
    }
    @Override
    public void onBindViewHolder(@NonNull RecycleAdapterUsers.ViewHolder holder, int position) {
        Participant participant = participants.get(position);
        SchoolInfo school = schoolsForUsers.get(position);
        holder.Name.setText(participant.getName() + " " + participant.getSurname());
        holder.Grade.setText(participant.getGrade().getName() + " класс");
        holder.SchoolName.setText(school.getName());
        holder.scan.setOnClickListener(v -> {
            NavDirections action =
                    WebLoginFragmentDirections.toWebLogin(participant.getSysGuid(),
                            participant.getName() + " " + participant.getSurname());
            Navigation.findNavController(fragment.requireActivity(), R.id.nav_host_fragment_content_main).navigate(action);
        });
        holder.share.setOnClickListener(v -> {
            NavDirections action =
                    ShareQRFragmentDirections.toShare(new ShareUser(participant.getName() + " " + participant.getSurname(),
                            participant.getSysGuid(), school.getShortName(), participant.getGrade().getName()));
            Navigation.findNavController(fragment.requireActivity(), R.id.nav_host_fragment_content_main).navigate(action);
        });
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                CommonAPI.getInstance().ChangeUuid(participant.getSysGuid());
            }
        });
    }

    @Override
    public int getItemCount() {
        return participants.size();
    }
    public static class ViewHolder extends RecyclerView.ViewHolder {

        TextView Grade;
        TextView Name;
        TextView SchoolName;
        ImageView scan;
        ImageView share;
        // Get the handles by calling findViewById() on View object inside the constructor
        ViewHolder(View v)
        {
            super(v);
            Grade = v.findViewById(R.id.user_Grade);
            Name = v.findViewById(R.id.user_Name);
            SchoolName = v.findViewById(R.id.user_SchoolName);
            scan = v.findViewById(R.id.scanQr);
            share = v.findViewById(R.id.shareQr);
        }
    }
}

File: app/src/main/java/com/alex/materialdiary/sys/adapters/RecycleAdapterUsers.kt
package com.alex.materialdiary.sys.adapters

import androidx.navigation.Navigation.findNavController
import com.alex.materialdiary.sys.net.PskoveduApi.Companion.getInstance
import com.alex.materialdiary.ChangeUserFragment
import com.alex.materialdiary.sys.net.models.get_user.Schools
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import com.alex.materialdiary.sys.net.models.get_user.SchoolInfo
import android.view.ViewGroup
import android.widget.ImageView
import com.alex.materialdiary.R
import com.alex.materialdiary.sys.net.models.ShareUser
import android.widget.TextView
import androidx.navigation.NavDirections
import com.alex.materialdiary.ShareQRFragmentDirections
import com.alex.materialdiary.sys.net.models.get_user.Participant
import java.util.ArrayList

class RecycleAdapterUsers(context: ChangeUserFragment, schools: List<Schools>?) :
    RecyclerView.Adapter<RecycleAdapterUsers.ViewHolder>() {
    private val inflater: LayoutInflater
    private val fragment: ChangeUserFragment
    private val participants: List<Participant?>?
    private val schoolsForUsers: MutableList<SchoolInfo?> = ArrayList()

    init {
        inflater = LayoutInflater.from(context.requireContext())
        participants = transform(schools)
        fragment = context
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = inflater.inflate(R.layout.users_item, parent, false)
        return ViewHolder(view)
    }

    fun transform(schools: List<Schools>?): List<Participant?>? {
        val list: MutableList<Participant?> = ArrayList()
        if (schools == null) return null
        for ((roles, school, _, _, _, participant, _, userParticipants) in schools) {
            if (roles.contains("participant")) {
                list.add(participant)
                schoolsForUsers.add(school)
            } else {
                list.addAll(userParticipants!!)
                for (i in userParticipants.indices) {
                    schoolsForUsers.add(school)
                }
            }
        }
        return list
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val participant = participants!![position]
        val school = schoolsForUsers[position]
        holder.Name.text = participant!!.name + " " + participant.surname
        holder.Grade.text = participant.grade!!.name + " класс"
        holder.SchoolName.text = school!!.name
        holder.share.setOnClickListener {
            val action: NavDirections = ShareQRFragmentDirections.toShare(
                ShareUser(
                    participant.name + " " + participant.surname,
                    participant.sysGuid!!, school.shortName!!, participant.grade!!.name!!
                )
            )
            findNavController(
                fragment.requireActivity(),
                R.id.nav_host_fragment_content_main
            ).navigate(action)
        }
        holder.itemView.setOnClickListener {
            val name = participant.surname + " " + participant.name + " " + participant.secondname
            getInstance(inflater.context).changeGuid(participant.sysGuid!!, name)
        }
    }

    override fun getItemCount(): Int {
        return participants!!.size
    }

    class ViewHolder internal constructor(v: View) : RecyclerView.ViewHolder(v) {
        var Grade: TextView
        var Name: TextView
        var SchoolName: TextView
        var scan: ImageView
        var share: ImageView

        // Get the handles by calling findViewById() on View object inside the constructor
        init {
            Grade = v.findViewById(R.id.user_Grade)
            Name = v.findViewById(R.id.user_Name)
            SchoolName = v.findViewById(R.id.user_SchoolName)
            scan = v.findViewById(R.id.scanQr)
            share = v.findViewById(R.id.shareQr)
        }
    }
}